### PR TITLE
Radiology tasks 6, 10, 11

### PR DESCRIPTION
**Pull request by Lena235711**

### CODE REVIEW

File: src/unicorn_eval/adaptors/__init__.py [hunk 1]
```diff
@@ -27,6 +27,7 @@
 )
 from unicorn_eval.adaptors.detection import DensityMap
 from unicorn_eval.adaptors.segmentation import SegmentationUpsampling
+from unicorn_eval.adaptors.segmentation_radiology import SegmentationUpsampling3D
```
Comment (clemsgrs) [inline src/unicorn_eval/adaptors/__init__.py:30]: move `SegmentationUpsampling3D` adaptor to `unicorn_eval.adaptors.segmentation`

File: src/unicorn_eval/adaptors/patch_extraction.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        world_coordinates = tuple(
+            image.TransformIndexToPhysicalPoint(coord)
+            for coord in matrix_coordinates
+        )
+        coordinates.append(world_coordinates)
+
+    return patches, coordinates
+
+
+if __name__ == '__main__':
```
Comment (clemsgrs) [inline src/unicorn_eval/adaptors/patch_extraction.py:96]: clean up this `if __name__ == '__main__'` code block

File: src/unicorn_eval/adaptors/segmentation_decoder.py
Comment (clemsgrs) [inline src/unicorn_eval/adaptors/segmentation_decoder.py:1]: move this code in `segmentation.py` file

File: src/unicorn_eval/adaptors/segmentation_radiology.py [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+    )
+
+    return patch_labels
+
+def load_patch_data(data_array: np.ndarray, batch_size: int = 4) -> DataLoader:
+    train_ds = Dataset(data=data_array)
+    train_loader = DataLoader(train_ds, batch_size=batch_size, shuffle=False)
+    return train_loader
+
+class SegmentationUpsampling3D(PatchLevelTaskAdaptor):
```
Comment (clemsgrs) [inline src/unicorn_eval/adaptors/segmentation_radiology.py:290]: can you briefly describe the different steps to go from features to predictions using the class docstring?

File: src/unicorn_eval/adaptors/segmentation_radiology.py [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+            image_sizes=self.test_image_sizes,
+            image_origins=self.test_image_origins,
+            image_spacings=self.test_image_spacings,
+            image_directions=self.test_image_directions,
+        )
+        test_loader = load_patch_data(test_data)
+
+        # run inference using the trained decoder
+        return inference(self.decoder, test_loader, self.device, self.return_binary)
+
```
Comment (clemsgrs) [inline src/unicorn_eval/adaptors/segmentation_radiology.py:394]: clean up empty lines

File: src/unicorn_eval/adaptors/segmentation_radiology.py
Comment (clemsgrs) [inline src/unicorn_eval/adaptors/segmentation_radiology.py:1]: move this code in `segmentation.py` file

File: src/unicorn_eval/evaluate.py [hunk 5]
```diff
@@ -33,17 +35,21 @@
     write_json_file,
 )
 
-INPUT_DIRECTORY = Path("/input")
-OUTPUT_DIRECTORY = Path("/output")
-GROUNDTRUTH_DIRECTORY = Path("/opt/ml/input/data/ground_truth")
+
+INPUT_DIRECTORY = Path("/data/temporary/unicorn/debugging/evaluation/vision/task-6")
```
Comment (clemsgrs) [inline src/unicorn_eval/evaluate.py:39]: please restore to previous value

File: src/unicorn_eval/evaluate.py [hunk 6]
```diff
@@ -33,17 +35,21 @@
     write_json_file,
 )
 
-INPUT_DIRECTORY = Path("/input")
-OUTPUT_DIRECTORY = Path("/output")
-GROUNDTRUTH_DIRECTORY = Path("/opt/ml/input/data/ground_truth")
+
+INPUT_DIRECTORY = Path("/data/temporary/unicorn/debugging/evaluation/vision/task-6")
+OUTPUT_DIRECTORY = Path("/data/bodyct/experiments/lena_t10027/unicorn_internal_dev/")
```
Comment (clemsgrs) [inline src/unicorn_eval/evaluate.py:40]: please restore to previous value

File: src/unicorn_eval/evaluate.py [hunk 7]
```diff
@@ -33,17 +35,21 @@
     write_json_file,
 )
 
-INPUT_DIRECTORY = Path("/input")
-OUTPUT_DIRECTORY = Path("/output")
-GROUNDTRUTH_DIRECTORY = Path("/opt/ml/input/data/ground_truth")
+
+INPUT_DIRECTORY = Path("/data/temporary/unicorn/debugging/evaluation/vision/task-6")
+OUTPUT_DIRECTORY = Path("/data/bodyct/experiments/lena_t10027/unicorn_internal_dev/")
+GROUNDTRUTH_DIRECTORY = Path("/data/temporary/unicorn/debugging/groundtruth/")
```
Comment (clemsgrs) [inline src/unicorn_eval/evaluate.py:41]: please restore to previous value

File: src/unicorn_eval/evaluate.py [hunk 8]
```diff
@@ -59,22 +65,36 @@
     "Task05_detecting_signet_ring_cells_in_he_stained_wsi_of_gastric_cancer": [
         "histopathology-region-of-interest-cropout"
     ],
+    "Task06_detecting_clinically_significant_prostate_cancer_in_mri_exams": [
+        "transverse-t2-prostate-mri"
```
Comment (clemsgrs) [inline src/unicorn_eval/evaluate.py:69]: is there a single input for task 6? i thought there were 3

File: src/unicorn_eval/evaluate.py [hunk 9]
```diff
@@ -83,8 +103,12 @@
     "Task03_predicting_the_time_to_biochemical_recurrence_in_he_prostatectomies": "overall-survival-years.json",
     "Task04_predicting_slide_level_tumor_proportion_score_in_ihc_stained_wsi": "pd-l1-tps-binned.json",
     "Task05_detecting_signet_ring_cells_in_he_stained_wsi_of_gastric_cancer": "cell-classification.json",
+    "Task06_detecting_clinically_significant_prostate_cancer_in_mri_exams": "images/transverse-cspca-label/{case_id}.mha", 
+
```
Comment (clemsgrs) [inline src/unicorn_eval/evaluate.py:107]: clean up empty line

File: src/unicorn_eval/evaluate.py [hunk 10]
```diff
@@ -355,14 +385,25 @@ def load_tif_file(*, location):
...
+    return class_labels
 
 def write_metrics(*, metrics):
     # write a json document used for ranking results on the leaderboard
     with open(OUTPUT_DIRECTORY / "metrics.json", "w") as f:
         f.write(json.dumps(metrics, indent=4))
 
 
-def write_combined_metrics(*, metric_dict: dict[dict]) -> None:
+def write_combined_metrics(*, metric_dict: dict[dict], json_only: bool = False) -> None:
```
Comment (clemsgrs) [inline src/unicorn_eval/evaluate.py:406]: could you use a more explicit variable name? probably `save_predictions: bool = False` is more adapted.

File: src/unicorn_eval/evaluate.py [hunk 11]
```diff
@@ -525,11 +574,18 @@ def main():
...
+            shot_image_directions = results["shot_image_directions"]
             case_embeddings = results["case_embeddings"]
 
             case_extra_labels = results["case_extra_labels"]
             patch_size = results["patch_size"]
             case_image_size = results["cases_image_sizes"]
+            case_image_spacings = results["cases_image_spacings"]
+            case_image_origins = results["cases_image_origins"]
+            case_image_directions = results["cases_image_directions"]
+
```
Comment (clemsgrs) [inline src/unicorn_eval/evaluate.py:588]: clean up empty line

File: src/unicorn_eval/evaluate.py [hunk 12]
```diff
@@ -573,11 +633,14 @@ def main():
             test_predictions=predictions,
             test_labels=case_labels,
             test_extra_labels=case_extra_labels,
+            test_image_spacing=case_image_spacings,
         )
         task_metrics[task_name] = metrics
 
-    write_combined_metrics(metric_dict=task_metrics)
-
+    if task_type == 'segmentation':
```
Comment (clemsgrs) [inline src/unicorn_eval/evaluate.py:640]: nice addition, could you use `"segmentation"` instead of `'segmentation'`
+ not forget to update the second argument with the more appropriated variable name chose (e.g. `save_prediction`)

File: src/unicorn_eval/utils.py
Comment (clemsgrs) [inline src/unicorn_eval/utils.py:1]: just to doublecheck: do you confirm the newly added metrics (`picai`, `uls`, `spider`) do range from 0 to 1 (and not e.g. from -1 to 1)?

File: src/unicorn_eval/utils.py [hunk 13]
```diff
@@ -223,7 +285,7 @@ def convert_numpy_types(obj):
         return obj
 
 
-def evaluate_predictions(task_name, case_ids, test_predictions, test_labels, test_extra_labels=None):
+def evaluate_predictions(task_name, case_ids, test_predictions, test_labels, test_extra_labels=None, test_image_spacing=None):
```
Comment (clemsgrs) [inline src/unicorn_eval/utils.py:288]: could you try to add the `test_image_spacing` to the `test_extra_labels` array? would be cleaner than having an argument that is only used in task 11

**Review by clemsgrs — COMMENTED** [review]
some clean up and a few fixes needed:
- could you briefly explain why you "split" the image label for task 6 into smaller patches? is the metric computed on a per-patch basis?
- why task6 inputs is only `transverse-t2-prostate-mri` and not the additional two MRIs?
- move adaptor segmentation code to the `segmentation.py` file

File: src/unicorn_eval/adaptors/segmentation_decoder.py
Comment (Lena235711) [reply src/unicorn_eval/adaptors/segmentation_decoder.py:1]: moved it
